agent/wsclient/client_test.go

```diff
@@ -0,0 +1,138 @@
+
+package wsclient
+
+import (
+	"io"
+	"os"
+	"testing"
+
+	"github.com/aws/amazon-ecs-agent/agent/acs/model/ecsacs"
+	"github.com/aws/amazon-ecs-agent/agent/config"
+	"github.com/aws/amazon-ecs-agent/agent/wsclient/mock/utils"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+
+	"github.com/gorilla/websocket"
+
+	"github.com/stretchr/testify/assert"
+)
+
+const dockerEndpoint = "/var/run/docker.sock"
+
+// TestConcurrentWritesDontPanic will force a panic in the websocket library if
+// the implemented methods don't handle concurrency correctly
+// See https://godoc.org/github.com/gorilla/websocket#hdr-Concurrency
+func TestConcurrentWritesDontPanic(t *testing.T) {
+	mockServer, _, requests, _, _ := mockwsutils.StartMockServer(t, make(chan []byte))
+	defer mockServer.Close()
+
+	req := ecsacs.AckRequest{Cluster: aws.String("test"), ContainerInstance: aws.String("test"), MessageId: aws.String("test")}
+
+	cs := getClientServer(mockServer.URL)
+	cs.Connect()
+
+	executeTenRequests := func() {
+		for i := 0; i < 10; i++ {
+			cs.MakeRequest(&req)
+		}
+	}
+
+	// Make requests from two separate routines to try and force a
+	// concurrent write
+	go executeTenRequests()
+	executeTenRequests()
+
+	t.Log("Waiting for all 20 requests to succeed")
+	for i := 0; i < 20; i++ {
+		<-requests
+	}
+}
+
+// TestProxyVariableCustomValue ensures that a user is able to override the
+// proxy variable by setting an environment variable
+func TestProxyVariableCustomValue(t *testing.T) {
+	mockServer, _, _, _, _ := mockwsutils.StartMockServer(t, make(chan []byte))
+	defer mockServer.Close()
+
+	testString := "Custom no proxy string"
+	os.Setenv("NO_PROXY", testString)
+	getClientServer(mockServer.URL).Connect()
+
+	assert.Equal(t, os.Getenv("NO_PROXY"), testString, "NO_PROXY should match user-supplied variable")
+}
+
+// TestProxyVariableDefaultValue verifies that NO_PROXY gets overridden if it
+// isn't already set
+func TestProxyVariableDefaultValue(t *testing.T) {
+	mockServer, _, _, _, _ := mockwsutils.StartMockServer(t, make(chan []byte))
+	defer mockServer.Close()
+
+	os.Unsetenv("NO_PROXY")
+	getClientServer(mockServer.URL).Connect()
+
+	expectedEnvVar := "169.254.169.254,169.254.170.2," + dockerEndpoint
+
+	assert.Equal(t, os.Getenv("NO_PROXY"), expectedEnvVar, "Variable NO_PROXY expected to be overwritten when no default value supplied")
+}
+
+// TestHandleMessagePermissibleCloseCode ensures that permissible close codes
+// are wrapped in io.EOF
+func TestHandleMessagePermissibleCloseCode(t *testing.T) {
+	closeWS := make(chan []byte)
+	messageError := make(chan error)
+	mockServer, _, _, _, _ := mockwsutils.StartMockServer(t, closeWS)
+	cs := getClientServer(mockServer.URL)
+	cs.Connect()
+
+	go func() {
+		messageError <- cs.ConsumeMessages()
+	}()
+
+	closeWS <- websocket.FormatCloseMessage(websocket.CloseNormalClosure, ":)")
+	assert.EqualError(t, <-messageError, io.EOF.Error(), "expected EOF for normal close code")
```

Comment:
It looks like it would! But it won't even compile if I try that. EqualError() wants a string. :man_shrugging: